### ISSUE TITLE
Impossible to use "curried" operators

### ISSUE THREAD

**Issue by gaku-sei** [rescript-lang/syntax#105]
In OCaml (and old ReasonML), this works:

```ocaml
let addOne = (+) 1 in

Js.log (addOne 41) (* Displays 42 *)
```

But it turns out the above doesn't work with the new syntax.
That being said, the following does compile:

```reason
let addOne = \"+"(1)

Js.log(addOne(41)) // Displays 42 as well
```

IMHO `\"+"(1)` is not easier to read, or clearer than `(+)` 😕 
Especially since theses operators will most of the time be read as strings which they are not.

 Is it by design, or is it in progress?

**Comment by chenglou** [comment]
Did you get this from `bsc -format`? Because that's just an automated best-effort translation. This isn't recommended at all.

We don't support infix in prefix position officially. The `\"+"(1)` happens to be an escape hatch only so that your conversion doesn't fail.

**Comment by gaku-sei** [comment]
Thank you @chenglou 😄 
I, indeed, used `bsc -format` 😕 

> We don't support infix in prefix position officially

Is is a planned feature? It's to be a common pattern 😕

The former syntax (using parentheses) is maybe not perfect, but we could at least drop the quotes...

Something like:

```reason
let sum = [1, 2, 3]->Belt.Array.reduce(\+)
```

Or even:

```reason
// No parentheses around the operator itself, might be hard to parse sometimes though
let sum = [1, 2, 3]->Belt.Array.reduce(+)

// As a terser equivalent to
let add = (x, y) => x + y

let sum = [1, 2, 3]->Belt.Array.reduce(add)
```

**Comment by chenglou** [comment]
It's unclear that we'll support this, though we definitely discourage such pattern. They don't save many keystrokes and usually obscure the meaning of the call (unless in trivial cases like + and *)
We recommend you to write `reduce((a, b) => a + b)` instead. Immediately clear, more refactorable, same perf (closure-less lambdas are lifted), and doesn't distract in terms of discourse. More importantly, avoid `reduce` when possible; carry around a `sum` helper here instead.

**Comment by gaku-sei** [comment]
Hum, that's a bit disappointed actually 😕 

> though we definitely discourage such pattern

Why? They _do_ save many keystrokes (in that case `(a, b) => a + b` becomes `(+)`, that is 5x less), and more importantly the latter function call makes the code extremely clear.

Don't get me wrong, I'm also into "easy to read" over "short to write" code in general, I just don't see how an inlined function will be clearer in that exact scenario, especially since it tends to add a lot of noise around the core logic: "+".

Does it mean passing a function is discouraged? If not, why, especially when the operator is a common as "+"?

Also, I'm not sure about this:

>More importantly, avoid reduce when possible; carry around a sum helper here instead

Doest it mean the `reduce` function is discouraged? Why so? Also, is it related to the currying by default? In that case, is it "ok" to use `reduceU`?

**Comment by chenglou** [comment]
> that is 5x less

Using a multiplier here isn't a great argument especially since the base amount is so tiny. There are more worthwhile bigger paradigms. In general we'd like to avoid "little abstractions". Also consider that we have collectively spent more time writing these sentences over the years than the total keystrokes saving ever, from this composition. (Note also the future progression of this thread btw)

> especially since it tends to add a lot of noise around the core logic: "+"

The core logic is rarely ever `+`. And the noise of `reduce` champions the lack of noise in the rest. This very specific conciseness doesn't really generalize.

The issue is that for these things the example is always a triviality like `+`. If it's that common then you should be using a `sum` which, if we take the 5x argument seriously, would be an even larger save on keystrokes (not really though). The only other good, somewhat-used example is `*` anyway.

> especially when the operator is a common as "+" 

In reality it's more like such `reduce(+)` case constitutes a tiny portion of one's code (though a large portion of one's attention in discussions) unless one shoehorned it and/or didn't have a straightforward `sum` helper.

> Doest it mean the `reduce` function is discouraged

Yes, `reduce` is discouraged. And no it has nothing to do with currying; more about readability (granted, the readability is mostly fine in this case with `(+)` if you've assimilated the execution of `reduce` already). And yes, in this case, one could argue that the readability is slightly affected, but in most other cases the readability of `reduce` is not higher.

As an aside, the unfortunate thing with some FP discussions is that many of these helpers, which happen to be functions, get lumped into the idea of _functional programming_, when in reality they're really just awkward helper functions. Not every function shed FP in a good light. If we can consider this from the perspective of "this is just an unfortunately overused helper function whether it's in OO or FP" then I think this'd make more sense.

Note that these are recommendations from experience. You're free to use and experiment with whatever FP concept you like for yourself, because I don't really believe words are enough to explain these. Though don't forget that our FP is at the service of shipping, not the other way around!

**Comment by gaku-sei** [comment]
Ok, let's go back to the original questions:

1. Is the lack of support of curried operators an unimplemented feature, or by design?
2. If it's by design, why?

To me something like `reduce(+)` is clearer than the version with the function inlined, not because of the keystrokes (which I don't care about), but because of the "noise" the inlined function adds, especially since, as you mentionned, real life code is way less trivial than summing arrays of ints.

You said that the inlined function didn't save many keystrokes, and I just replied it is not _in this exact case_. And more importantly, as already stated: keystrokes are not the point.

I think an other question could be, why is `reduce(add)` allowed when `reduce(+)` is banned?

To be clear, I'm not stating the latter is better, just trying to understand why higher order functions are ok, when "higher ordered operators" are not (and that was implicitly part of the original question).

To give a bit more background to the question, I have some codebases that are getting pretty large, mostly in OCaml/ReasonML on the frontend, and I do use operators, higher ordered operators a lot, functions composition, etc... (Whether we agree on how useful, or _good_ these techniques are, is an other story). And the new syntax "breaks" a lot of code.

Again, not trying to have an "old vs new syntax" conversation here, I have my opinions and you have yours, I just wanted to clarify something that looked odd to me. Especially since I don't have the choice but to move to the new syntax (I don't buy the "we'll keep supporting OCaml/ReasonML").

>Note that these are recommendations from experience. You're free to use and experiment with whatever FP concept you like for yourself

Thank you for allowing different opinions.

**Comment by TheSpyder** [comment]
> You're free to use and experiment with whatever FP concept you like for yourself

Except when language decisions prevent us from doing so. If `reduce` is removed from the standard library, we can write our own, but we're at the mercy of the parser when it comes to custom infix and currying patterns.

I'm in a similar boat to @gaku-sei; I'm working in a team that would be quite comfortable with higher-order operators if a case to use one comes up, just like we're comfortable with infix patterns that look like line noise if you're not used to them.

**Comment by jfrolich** [comment]
I agree that some functional programming concepts can be hard to read especially for newcomers. But I've never felt that `reduce` or `foldLeft` is that way. It's pretty intuitive for me and it helps that there is no mutation happening outside of the function that I have to worry about. But it can be different for different persons. I think we also don't have to let the pendulum swing the other way, that mutation + imperative constructs are preferred, the reason this feels more intuitive to some is also because of the baggage of years of writing in an imperative style (it's not an objective quality that the other notation is more clear).
